example/client/cmd/util.go

```diff
@@ -50,25 +54,124 @@ func deleteAPIClient(cs *clientState) {
 	cs.connection = nil
 }
 
-func parseAttributes(attributes string) (map[string]string, error) {
-	attrs := make(map[string]string)
-	if len(attributes) > 0 {
-		for _, a := range strings.Split(attributes, ",") {
-			i := strings.Index(a, "=")
-			if i < 0 {
-				return nil, fmt.Errorf("Attribute value %v does not include an = sign", a)
-			} else if i == 0 {
-				return nil, fmt.Errorf("Attribute value %v does not contain a valid name", a)
+func parseAttributes(rootArgs *rootArgs) (*mixerpb.Attributes, error) {
+	attrs := mixerpb.Attributes{}
+	attrs.Dictionary = make(map[int32]string)
+
+	// once again, the following boilerplate would be more succinct with generics...
+
+	type AttrType int
+	const (
+		STRING AttrType = iota
+		INT64
+		DOUBLE
+		BOOL
+		TIMESTAMP
+		BYTES
+	)
+
+	for i := STRING; i <= BYTES; i++ {
+		var a string
+		switch i {
+		case STRING:
+			a = rootArgs.stringAttributes
+		case INT64:
+			a = rootArgs.int64Attributes
+		case DOUBLE:
+			a = rootArgs.doubleAttributes
+		case BOOL:
+			a = rootArgs.boolAttributes
+		case TIMESTAMP:
+			a = rootArgs.timestampAttributes
+		case BYTES:
+			a = rootArgs.boolAttributes
+		}
+
+		if len(a) > 0 {
+			for _, a := range strings.Split(a, ",") {
+				eq := strings.Index(a, "=")
+				if eq < 0 {
+					return nil, fmt.Errorf("Attribute value %v does not include an = sign", a)
+				}
+				if eq == 0 {
+					return nil, fmt.Errorf("Attribute value %v does not contain a valid name", a)
+				}
+				name := a[0:eq]
+				value := a[eq+1:]
+
+				index := int32(len(attrs.Dictionary))
+				attrs.Dictionary[index] = name
+
+				switch i {
```

Comment:
can we eliminate this ugliness with something more broken down into components and using code like:

```

func attrMap(s string) (map[string]string, error) {
	m := make(map[string]string)
	for _, attrStr := range strings.Split(s, ",") {
		keyVal := strings.Split(attrStr, "=")
		if len(keyVal) != 2 || keyVal[0] == "" || keyVal[1] == "" {
			return nil, errors.New("doh!")
		}
		m[keyVal[0]] = keyVal[1]
	}
	return m, nil
}

type convertFn func(string) (interface{}, error)

var (
	int64Fn = func(s string) (interface{}, error) { return strconv.ParseInt(s, 10, 64) }
	boolFn = func(s string) (interface{}, error) { return strconv.ParseBool(s) }
)

func convert(m map[string]string, f convertFn) (map[string]interface{}, error) {
	n := make(map[string]interface{}, len(m))
	for k, v := range m {
		nv, err := f(v)
		if err != nil {
			return nil, err
		}
		n[k] = nv
	}
	return n, nil
}

func toBoolMap(m map[string]interface{}) (map[string]bool, error) {
	out := make(map[string]bool, len(m))
	for k, v := range m {
		out[k] = v.(bool)
	}
	return out, nil
}


func main() {

	var good = "t=false,s=3453,z=testing"
	fmt.Println(attrMap(good))

	var bad = "t=134,s=234234,=234,w="
	fmt.Println(attrMap(bad))
	
	i, _ := attrMap("t=false,s=true")
	fmt.Println(convert(i, boolFn))
	
	c, _ := convert(i, boolFn)
	fmt.Println(toBoolMap(c))
}
```

The advantages would be smaller, more legible (debateable perhaps) code.